app/gui2/src/util/containers.ts

```diff
@@ -0,0 +1,246 @@
+import { assertDefined, assertEqual } from '@/util/assert'
+import { expect, test } from 'vitest'
+
+/**
+ * Map that supports Object-based keys.
+ *
+ * Internally keys are converted to strings using the provided {@link keyMapper} function and then compared.
+ *
+ * @template Key The type of the keys.
+ * @template Value The type of the values.
+ */
+export class ObjectKeyedMap<Key extends Object, Value> {
```

Comment:
That class name is a bit confusing, because standard `Map` also supports being keyed by any objects. So it is not clear what it really does beyond that. The issue with standard map it is that it uses object's identity as key (matching `===` operator semantics). Two identical, but separate object instances are effectively two unique hashmap keys. It would be good to convey that behavior difference in this class name.

Also, is would be cool to relax the `keyMapper` return type from a string to any valid sensible map key. Also, `JSON.stringify` will also give you sometimes surprising results. Those two objects will be considered different map keys:
```ts
const x = { a: 1, b: 1 }
const y = { b: 1, a: 1 }

map.set(x, true)
map.get(y) // -> undefined
```

Reply:
Call notes:
use `keyForRange`
don't use JSON.stringify as default

Reply:
Applied changes, as discussed.